impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -150,11 +129,8 @@ private KafkaConfig buildBrokerConfig(KafkaClusterConfig.ConfigHolder c, Path te
     public synchronized void start() {
         // kraft mode: per-broker: 1 external port + 1 inter-broker port + 1 controller port + 1 anon port
         // zk mode: per-cluster: 1 zk port; per-broker: 1 external port + 1 inter-broker port + 1 anon port
-        try (var preallocator = new ListeningSocketPreallocator()) {
-            List.of(Listener.EXTERNAL, Listener.ANON, Listener.INTERNAL)
-                    .forEach(l -> Utils.putAllListEntriesIntoMapKeyedByIndex(preallocator.preAllocateListeningSockets(clusterConfig.getBrokersNum()), ports.get(l)));
-            Utils.putAllListEntriesIntoMapKeyedByIndex(allocateControllerPorts(preallocator, clusterConfig), ports.get(Listener.CONTROLLER));
-        }
+        portsAllocator.allocate(Set.of(Listener.EXTERNAL, Listener.ANON, Listener.INTERNAL), 0, clusterConfig.getBrokersNum());
+        portsAllocator.allocate(Set.of(Listener.CONTROLLER), 0, clusterConfig.isKraftMode() ? clusterConfig.getKraftControllers() : 1);
```

Comment:
I've realised this is a problem, since there is a new `ListeningSocketPreallocator` created per `allocate` we might allocate the same port twice.

Reply:
I have introduced the idea of an `allocationSession` to work around this, so the same preallocator is used for both calls.